lib/Transforms/ConvertSecretExtractToStaticExtract/ConvertSecretExtractToStaticExtract.cpp

```diff
@@ -0,0 +1,212 @@
+
+struct SecretExtractToStaticExtractConversion
+    : OpRewritePattern<tensor::ExtractOp> {
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+
+ public:
+  SecretExtractToStaticExtractConversion(DataFlowSolver *solver,
+                                         MLIRContext *context)
+      : OpRewritePattern(context), solver(solver) {}
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp extractOp,
+                                PatternRewriter &rewriter) const override {
+    // TODO(#866): Add support for reads on multi-dimensional tensors
+    if (extractOp.getTensor().getType().getRank() != 1) {
+      extractOp->emitWarning()
+          << "Currently, transformation only supports 1D tensors:";
+      return failure();
+    }
+
+    auto index = extractOp.getIndices().front();
+
+    auto *indexSecretnessLattice =
+        solver->lookupState<SecretnessLattice>(index);
+
+    // If (1) lattice is not found or (2) lattice is set to unknown, apply
+    // transformation
+    if (!indexSecretnessLattice ||
+        !indexSecretnessLattice->getValue().isInitialized()) {
+      extractOp->emitWarning()
+          << "Secretness for tensor.extract index has not been set. "
+             "Conservatively, the transformation will be applied:";
+    }
+
+    auto indexSecretness = indexSecretnessLattice->getValue();
+
+    bool isIndexSecret =
+        !indexSecretness.isInitialized() ||
+        (indexSecretness.isInitialized() && indexSecretness.getSecretness());
+
+    // If index is not secret, no transformation is needed
+    if (!isIndexSecret) return failure();
+
+    ImplicitLocOpBuilder builder(extractOp->getLoc(), rewriter);
+
+    auto argType = extractOp.getResult().getType();
+
+    auto *extractOpLattice =
+        solver->lookupState<SecretnessLattice>(extractOp.getResult());
+
+    auto extractOpSecretness = extractOpLattice ? extractOpLattice->getValue() :
+                                                /*unknown*/ Secretness();
+
+    // Placeholder for extracted value
+    auto dummyValue = builder.create<arith::ConstantOp>(
+        argType, builder.getIntegerAttr(argType, 0));
+    // Set secretness for dummyValue
+    setValueToSecretness(solver, dummyValue, extractOpSecretness);
+
+    int size = extractOp.getTensor().getType().getShape().front();
+
+    SmallVector<Value> iterArgs = {dummyValue};
+    auto forOp = builder.create<affine::AffineForOp>(0, size, 1, iterArgs);
+    // Set secretness for induction variable
+    setValueToSecretness(solver, forOp.getInductionVar(), Secretness(false));
+
+    builder.setInsertionPointToStart(forOp.getBody());
+
+    // Check if the current index is equal to the secret index
+    auto cond = builder.create<arith::CmpIOp>(arith::CmpIPredicate::eq,
+                                              forOp.getInductionVar(), index);
+    setSecretnessOfResults(solver, cond->getOperands(), cond->getResults());
```

Comment:
We know that the new `arith.cmpi` compares a public variable (the induction variable) with the secret-or-unknown index, so we can just directly set the secretness to `indexSecretness` via `setValueToSecretness`